### ISSUE TITLE
cmd/go: expose InvalidGoFiles as part of `go list` API

### ISSUE THREAD

**Issue by stamblerre** [golang/go#39986]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version devel +a4ba411b19 Wed Jul 1 15:29:29 2020 +0000 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### Repro steps

Consider a directory structure like the following:

```
mod2
├── go.mod
├── go.sum
├── inner
│   └── inner.go
└── main.go

1 directory, 4 files
```

If `inner.go` and `main.go` are both empty files, `go list` returns the following:

```
$ go list -e -json -compiled -test ./...
{
	"Dir": "/Users/rstambler/modules/mod2",
	"ImportPath": "mod2",
	"Root": "/Users/rstambler/modules/mod2",
	"Match": [
		"./..."
	],
	"Incomplete": true,
	"Error": {
		"ImportStack": [],
		"Pos": "main.go:1:1",
		"Err": "expected 'package', found 'EOF'"
	}
}
{
	"Dir": "/Users/rstambler/modules/mod2/inner",
	"ImportPath": "mod2/inner",
	"Root": "/Users/rstambler/modules/mod2",
	"Match": [
		"./..."
	],
	"Incomplete": true,
	"Error": {
		"ImportStack": [],
		"Pos": "inner/inner.go:1:1",
		"Err": "expected 'package', found 'EOF'"
	}
}
```

To extract the Go files associated with these packages, go/packages will have to parse the error messages.
Is there any possibility of listing the empty files in the GoFiles field?

/cc @jayconrod @bcmills @matloob

**Comment by jayconrod** [comment]
I think we could do this. `go/build.Context.ImportDir` a package with these files `InvalidGoFiles`, along with an error. The go command sees that error and ignores the returned package. If the package is non-nil though, we could copy a few fields like this where it makes sense.

**Comment by stamblerre** [comment]
We've encountered some limitations with these CLs - specifically with this corner case:

`hello.go` and `hello_test.go` are both empty on disk, but in overlays, their contents are `package hello` and `package hello_test`. In this case, `go list` will report both of these files as part of `hello`. The result of this is that a go/packages contains query will report `hello_test`'s package as the test variant of `hello`, not as an x test. This results in a lot of complexity for `go/packages` to work-around, and the goal of this was to avoid such work-arounds, so it doesn't seem worth it.

As an alternative, we proposed exposing the InvalidFiles list from go/build as part of the `go list` API. This would allow `go/packages` to treat the invalid files separately from a package, and then create the correct packages based on the overlays.

/cc @matloob
